common/dal2/src/ops/io.rs

```diff
@@ -117,3 +125,110 @@ where F: FnMut(usize)
         r
     }
 }
+/// Create a buffered `SeekableReader`
+///
+/// ## Note
+///
+/// We will use 4MB buffer for now to avoid too much repeated requests.
+#[allow(dead_code)]
+pub async fn new_buffered_seekable_reader<'d, S>(
+    da: Arc<crate::DataAccessor<'d, S>>,
+    key: &str,
+) -> Result<BufReader<SeekableReader<'d, S>>>
+where
+    S: super::Read<S> + super::Stat<S>,
+{
+    let o = da.stat(key).await?;
+
+    // We will create a BufReader with a large buf so it only read big chunk from underlying storage.
+    Ok(BufReader::with_capacity(
+        4 * 1024 * 1024, // 4 MiB
+        SeekableReader::new(da, key, o.size),
+    ))
+}
+
+/// If we already know a file's total size, we can implement Seek for it.
+///
+/// - Every time we call `read` we will send a new http request to fetch data from cloud storage like s3.
+/// - Every time we call `seek` we will update the `pos` field just in memory.
+/// # NOTE
+///
+/// It's better to use SeekableReader as an inner reader inside BufReader.
+///
+/// # TODO
+///
+/// We need use update the metrics.
+pub struct SeekableReader<'d, S: super::Read<S>> {
+    da: Arc<crate::DataAccessor<'d, S>>,
+    key: String,
+    total: u64,
+
+    pos: u64,
+}
+
+impl<'d, S> SeekableReader<'d, S>
+where S: super::Read<S>
+{
+    pub fn new(da: Arc<crate::DataAccessor<'d, S>>, key: &str, total: u64) -> Self {
+        SeekableReader {
+            da,
+            key: key.to_string(),
+            total,
+
+            pos: 0,
+        }
+    }
+}
+
+impl<S> AsyncRead for SeekableReader<'_, S>
+where S: super::Read<S>
+{
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let key = self.key.clone();
+        let da = self.da.clone();
+        let pos = self.pos;
+
+        let n = block_on(async {
```

Comment:
are there any other workarounds other than `block_on`?

Reply:
I tried a lot in the past three days, and I can't find a good alternative for `block_on`.

Basicly, it seems that calling `reader.poll(cx)` inside `poll_read()` will lead to dead Loop, and I can't figure where and how to resolve it.

Here is a record of my blood and tears: https://note.xuanwo.io/#/page/jan%2020th%2C%202022

Reply:
>  it seems that calling reader.poll(cx) inside poll_read() will lead to dead Loop
 
 may I take a look at the code of this version?

Reply:
Please take a look at https://github.com/Xuanwo/databend/commits/dal2-apply-2, thanks in advance!

Reply:
After quick browsing of https://github.com/Xuanwo/databend/commits/dal2-apply-2 :

Instead of `dead loop`,  more likely,  the `Future` returned by `poll_read` might be kept/trapped in the pending state

https://github.com/Xuanwo/databend/blob/3ff525a704b0752ea8c8ebe4d7daad7bf2f6b5d6/common/dal2/src/ops/io.rs#L196-L210
~~~.rust
        let n = async {
            let mut builder = da.read(key.as_str());
            let r = builder
                .offset(pos)
                .size(buf.len() as u64)
                .run()
                .await
                .map_err(io::Error::other)?;


            Pin::new(r).read(buf).await
        };


        pin_mut!(n);


        let n = ready!(n.poll(cx))?;
~~~

IMO, `AsyncReader` should hold a reference to the n (the `Future`, not the `usize`), to keep it from being dropped (too early)

Reply:
@dantengsky Thanks for the great review! I managed to learn how rust future works, and come out a better solution for implementing the seek_reader. PTAL.